packages/go_router/lib/src/go_route_information_provider.dart

```diff
@@ -0,0 +1,120 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/services.dart';
+import 'package:flutter/widgets.dart';
+import 'package:go_router/src/go_route_information_parser.dart';
+
+/// The route information provider created by go_router
+class GoRouteInformationProvider extends RouteInformationProvider
+    with WidgetsBindingObserver, ChangeNotifier {
+  /// Creates a [GoRouteInformationProvider].
+  GoRouteInformationProvider({
+    required RouteInformation initialRouteInformation,
+    Listenable? refreshListenable,
+  })  : _refreshListenable = refreshListenable,
+        _value = initialRouteInformation {
+    _refreshListenable?.addListener(notifyListeners);
+  }
+
+  final Listenable? _refreshListenable;
+
+  // ignore: unnecessary_non_null_assertion
+  static WidgetsBinding get _binding => WidgetsBinding.instance;
+
+  @override
+  void routerReportsNewRouteInformation(RouteInformation routeInformation,
+      {RouteInformationReportingType type =
+          RouteInformationReportingType.none}) {
+    // Avoid adding a new history entry if the route is the same as before.
+    final bool replace = type == RouteInformationReportingType.neglect ||
+        (type == RouteInformationReportingType.none &&
+            _valueInEngine.location == routeInformation.location);
+    SystemNavigator.selectMultiEntryHistory();
+    // TODO(chunhtai): should report extra to to browser through state if
+    // possible.
+    SystemNavigator.routeInformationUpdated(
+      location: routeInformation.location!,
+      replace: replace,
+    );
+    _value = routeInformation;
+    _valueInEngine = routeInformation;
+  }
+
+  @override
+  RouteInformation get value => DebugGoRouteInformation(
+        location: _value.location,
+        state: _value.state,
+      );
+  RouteInformation _value;
+  set value(RouteInformation other) {
+    final bool shouldNotify =
+        value.location != other.location || value.state != other.state;
```

Comment:
Should we compare `other` with `_value` instead of `value` to avoid allocating `DebugGoRouteInformation`?

```suggestion
        _value.location != other.location || _value.state != other.state;
```

Reply:
Good catch, no need to use public getter here.